File: src/codility/java/NumberOfDiscIntersections.java
package codility.java;

import java.util.Arrays;

/*
 * 92 m, find python coded solution 81% 
 * (because no proper long cast implemented)
 * and I missed task about > 10 000 000 return -1
 */
public class NumberOfDiscIntersections {
	
	public class Event implements Comparable<Event>{
		long coordinate;
		boolean begin;
		@Override
		public int compareTo(Event e) {
			if (this.coordinate < e.coordinate){
				return -1;
			} else if (this.coordinate == e.coordinate) {
				if (this.begin==true) return -1;
				else return 1;
			}else return 1;
		}
	}
	
	public int solution(int[] A) {
		Event[] events = new Event[A.length*2];
		for (int i=0; i<A.length; i++){
			Event event = new Event();
			event.coordinate = ((long)i-A[i]);
			event.begin=true;
			events[2*i] = event;
			event = new Event();
			event.coordinate = ((long)i+A[i]);
			event.begin=false;
			events[2*i+1] = event;
		}
		
		Arrays.sort(events);
		
		int intersections=0;
		int activeCircles=0;
		for (int i=0; i<events.length; i++){
			int countDelta = (events[i].begin==true) ? 1 : -1;
			if (countDelta>0){
				intersections += activeCircles*countDelta;				
			}
			activeCircles += countDelta;
			if (intersections > 10000000) return -1;
		}
		
		return intersections;
    }
}

File: src/codility/python/NumberOfDiscIntersections.py
def solution(A):
    events = []
    for i, a in enumerate(A):
        events += [(i-a, +1), (i+a, -1)]
    events.sort(key=lambda x: (x[0], -x[1]))
    intersections, active_circles = 0, 0
    for _, circle_count_delta in events:
        intersections += active_circles * (circle_count_delta > 0)
        active_circles += circle_count_delta
        if intersections > 10E6:
            return -1
    return intersections

if __name__ == '__main__':
    print ('Start tests..')
    assert solution([1, 5, 2, 1, 4, 0]) == 11
    assert solution([]) == 0
    assert solution([0,1]) == 1
    assert solution([0, 0]) == 0
    assert solution([1,0,0,3]) == 4
    print ('Ok!')